In [1]:
import pandas as pd
import numpy as np
import matplotlib as plt
import seaborn as sns

In [2]:
%matplotlib inline
pd.set_option('display.max_columns', None)

In [3]:
# import excel file Nov 2018 General results.xlsx
tn_2018g=pd.read_excel('../data/2018_election/Nov 2018 General results.xlsx',
                      usecols=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE','RNAME1','PARTY1','PVTALLY1',
                             'RNAME2','PARTY2','PVTALLY2','RNAME3','PARTY3','PVTALLY3','RNAME4','PARTY4','PVTALLY4',
                             'RNAME5','PARTY5','PVTALLY5','RNAME6','PARTY6','PVTALLY6','RNAME7','PARTY7','PVTALLY7',
                             'RNAME8','PARTY8','PVTALLY8','RNAME9','PARTY9','PVTALLY9','RNAME10','PARTY10','PVTALLY10'])
tn_2018g

COUNTY       PRECINCT  \
0      Anderson  Andersonville   
1      Anderson  Andersonville   
2      Anderson  Andersonville   
3      Anderson  Andersonville   
4      Anderson  Andersonville   
...         ...            ...   
13237    Wilson    Provisional   
13238    Wilson    Provisional   
13239    Wilson    Provisional   
13240    Wilson    Provisional   
13241    Wilson    Provisional   

                                              OFFICENAME   ELECTDATE  \
0                                               Governor  2018-11-06   
1                                               Governor  2018-11-06   
2                                               Governor  2018-11-06   
3                                   United States Senate  2018-11-06   
4      United States House of Representatives District 3  2018-11-06   
...                                                  ...         ...   
13237                               United States Senate  2018-11-06   
13238  United States House of Representatives District 6  2018-11-06   
13239                       Tennessee Senate District 17  2018-11-06   
13240     Tennessee House of Representatives District 46  2018-11-06   
13241     Tennessee House of Representatives District 57  2018-11-06   

                       RNAME1       PARTY1  PVTALLY1              RNAME2  \
0                    Bill Lee   Republican      1011           Karl Dean   
1          Tommy Ray McAnally  Independent         0  Jessie D. McDonald   
2      Tracy C. Yaste Tisdale  Independent         0          Mike Toews   
3            Marsha Blackburn   Republican       954       Phil Bredesen   
4           Chuck Fleischmann   Republican      1010   Danielle Mitchell   
...                       ...          ...       ...                 ...   
13237        Marsha Blackburn   Republican        54       Phil Bredesen   
13238               John Rose   Republican        54         Dawn Barlow   
13239               Mark Pody   Republican        61    Mary Alice Carfi   
13240              Clark Boyd   Republican        18          Mark Cagle   
13241            Susan M Lynn   Republican        37         Jordan Cole   

            PARTY2  PVTALLY2                  RNAME3       PARTY3  PVTALLY3  \
0       Democratic     244.0    Mark CoonRippy Brown  Independent       2.0   
1      Independent       0.0  Toney Randall Mitchell  Independent       0.0   
2      Independent       0.0              Rick Tyler  Independent       0.0   
3       Democratic     289.0         Trudy A. Austin  Independent       5.0   
4       Democratic     215.0              Rick Tyler  Independent      16.0   
...            ...       ...                     ...          ...       ...   
13237   Democratic      35.0         Trudy A. Austin  Independent       0.0   
13238   Democratic      30.0              Lloyd Dunn  Independent       3.0   
13239   Democratic      28.0                     NaN          NaN       NaN   
13240   Democratic      10.0                     NaN          NaN       NaN   
13241   Democratic      21.0                     NaN          NaN       NaN   

                RNAME4       PARTY4  PVTALLY4               RNAME5  \
0      Sherry L. Clark  Independent       4.0       Justin Cornett   
1       Yvonne Neubert  Independent       0.0  Alfred Shawn Rapoza   
2      Vinnie Vineyard  Independent       0.0     Jaron D. Weidner   
3          John Carico  Independent       5.0            Dean Hill   
4                  NaN          NaN       NaN                  NaN   
...                ...          ...       ...                  ...   
13237      John Carico  Independent       1.0            Dean Hill   
13238       David Ross  Independent       2.0                  NaN   
13239              NaN          NaN       NaN                  NaN   
13240              NaN          NaN       NaN                  NaN   
13241              NaN          NaN       NaN                  NaN   

            PARTY5  PVTALLY5    

In [4]:
tn_2018g.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13242 entries, 0 to 13241
Data columns (total 34 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   COUNTY      13242 non-null  object 
 1   PRECINCT    13242 non-null  object 
 2   OFFICENAME  13242 non-null  object 
 3   ELECTDATE   13242 non-null  object 
 4   RNAME1      13242 non-null  object 
 5   PARTY1      13242 non-null  object 
 6   PVTALLY1    13242 non-null  int64  
 7   RNAME2      12606 non-null  object 
 8   PARTY2      12483 non-null  object 
 9   PVTALLY2    12606 non-null  float64
 10  RNAME3      10363 non-null  object 
 11  PARTY3      10103 non-null  object 
 12  PVTALLY3    10363 non-null  float64
 13  RNAME4      8967 non-null   object 
 14  PARTY4      8677 non-null   object 
 15  PVTALLY4    8967 non-null   float64
 16  RNAME5      8130 non-null   object 
 17  PARTY5      8130 non-null   object 
 18  PVTALLY5    8130 non-null   float64
 19  RNAME6      8130 non-null

In [5]:
# merge entries in all RNAME columns into new column "name"
tn_2018g_namemelt=tn_2018g.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['RNAME1','RNAME2','RNAME3','RNAME4','RNAME5','RNAME6','RNAME7','RNAME8','RNAME9','RNAME10'],value_name="name")
# drop column "variable" created by .melt
tn_2018g_namemelt=tn_2018g_namemelt.drop(['variable'],axis=1)                          
tn_2018g_namemelt.head(9405)

COUNTY       PRECINCT  \
0     Anderson  Andersonville   
1     Anderson  Andersonville   
2     Anderson  Andersonville   
3     Anderson  Andersonville   
4     Anderson  Andersonville   
...        ...            ...   
9400     Roane      Lawnville   
9401     Roane         Renfro   
9402     Roane         Renfro   
9403     Roane         Renfro   
9404     Roane         Renfro   

                                             OFFICENAME   ELECTDATE  \
0                                              Governor  2018-11-06   
1                                              Governor  2018-11-06   
2                                              Governor  2018-11-06   
3                                  United States Senate  2018-11-06   
4     United States House of Representatives District 3  2018-11-06   
...                                                 ...         ...   
9400     Tennessee House of Representatives District 32  2018-11-06   
9401                                           Governor  2018-11-06   
9402                                           Governor  2018-11-06   
9403                                           Governor  2018-11-06   
9404                               United States Senate  2018-11-06   

                        name  
0                   Bill Lee  
1         Tommy Ray McAnally  
2     Tracy C. Yaste Tisdale  
3           Marsha Blackburn  
4          Chuck Fleischmann  
...                      ...  
9400             Kent Calfee  
9401                Bill Lee  
9402      Tommy Ray McAnally  
9403  Tracy C. Yaste Tisdale  
9404        Marsha Blackburn  

[9405 rows x 5 columns]

In [6]:
# merge entries in all PARTY columns into new column "party"
tn_2018g_partymelt=tn_2018g.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['PARTY1','PARTY2','PARTY3','PARTY4','PARTY5','PARTY6','PARTY7','PARTY8','PARTY9','PARTY10'],value_name="party")
# drop column "variable" created by .melt
tn_2018g_partymelt=tn_2018g_partymelt.drop(['variable'],axis=1)
tn_2018g_partymelt.head(9405)

COUNTY       PRECINCT  \
0     Anderson  Andersonville   
1     Anderson  Andersonville   
2     Anderson  Andersonville   
3     Anderson  Andersonville   
4     Anderson  Andersonville   
...        ...            ...   
9400     Roane      Lawnville   
9401     Roane         Renfro   
9402     Roane         Renfro   
9403     Roane         Renfro   
9404     Roane         Renfro   

                                             OFFICENAME   ELECTDATE  \
0                                              Governor  2018-11-06   
1                                              Governor  2018-11-06   
2                                              Governor  2018-11-06   
3                                  United States Senate  2018-11-06   
4     United States House of Representatives District 3  2018-11-06   
...                                                 ...         ...   
9400     Tennessee House of Representatives District 32  2018-11-06   
9401                                           Governor  2018-11-06   
9402                                           Governor  2018-11-06   
9403                                           Governor  2018-11-06   
9404                               United States Senate  2018-11-06   

            party  
0      Republican  
1     Independent  
2     Independent  
3      Republican  
4      Republican  
...           ...  
9400   Republican  
9401   Republican  
9402  Independent  
9403  Independent  
9404   Republican  

[9405 rows x 5 columns]

In [7]:
# merge entries in all PVTALLY columns into new column "vote_count"
tn_2018g_pvtallymelt=tn_2018g.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['PVTALLY1','PVTALLY2','PVTALLY3','PVTALLY4','PVTALLY5','PVTALLY6','PVTALLY7','PVTALLY8','PVTALLY9','PVTALLY10'],value_name="vote_count")
# drop column "variable" created by .melt
tn_2018g_pvtallymelt=tn_2018g_pvtallymelt.drop(['variable'],axis=1)
tn_2018g_pvtallymelt.head(9405)

COUNTY       PRECINCT  \
0     Anderson  Andersonville   
1     Anderson  Andersonville   
2     Anderson  Andersonville   
3     Anderson  Andersonville   
4     Anderson  Andersonville   
...        ...            ...   
9400     Roane      Lawnville   
9401     Roane         Renfro   
9402     Roane         Renfro   
9403     Roane         Renfro   
9404     Roane         Renfro   

                                             OFFICENAME   ELECTDATE  \
0                                              Governor  2018-11-06   
1                                              Governor  2018-11-06   
2                                              Governor  2018-11-06   
3                                  United States Senate  2018-11-06   
4     United States House of Representatives District 3  2018-11-06   
...                                                 ...         ...   
9400     Tennessee House of Representatives District 32  2018-11-06   
9401                                           Governor  2018-11-06   
9402                                           Governor  2018-11-06   
9403                                           Governor  2018-11-06   
9404                               United States Senate  2018-11-06   

      vote_count  
0         1011.0  
1            0.0  
2            0.0  
3          954.0  
4         1010.0  
...          ...  
9400       995.0  
9401       488.0  
9402         0.0  
9403         0.0  
9404       466.0  

[9405 rows x 5 columns]

In [8]:
# add "party" and "vote_count" columns to tn_2016_namemelt
party=tn_2018g_partymelt['party']
pvtally=tn_2018g_pvtallymelt['vote_count']
tn_2018g_namemelt=tn_2018g_namemelt.join(party).join(pvtally)
tn_2018g_namemelt.head(3)

COUNTY       PRECINCT OFFICENAME   ELECTDATE                    name  \
0  Anderson  Andersonville   Governor  2018-11-06                Bill Lee   
1  Anderson  Andersonville   Governor  2018-11-06      Tommy Ray McAnally   
2  Anderson  Andersonville   Governor  2018-11-06  Tracy C. Yaste Tisdale   

         party  vote_count  
0   Republican      1011.0  
1  Independent         0.0  
2  Independent         0.0

In [9]:
tn_2018g_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132420 entries, 0 to 132419
Data columns (total 7 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   COUNTY      132420 non-null  object 
 1   PRECINCT    132420 non-null  object 
 2   OFFICENAME  132420 non-null  object 
 3   ELECTDATE   132420 non-null  object 
 4   name        87282 non-null   object 
 5   party       84621 non-null   object 
 6   vote_count  87282 non-null   float64
dtypes: float64(1), object(6)
memory usage: 7.1+ MB


In [10]:
# lower case column names
tn_2018g_namemelt.columns=tn_2018g_namemelt.columns.str.lower()
tn_2018g_namemelt

county       precinct  \
0       Anderson  Andersonville   
1       Anderson  Andersonville   
2       Anderson  Andersonville   
3       Anderson  Andersonville   
4       Anderson  Andersonville   
...          ...            ...   
132415    Wilson    Provisional   
132416    Wilson    Provisional   
132417    Wilson    Provisional   
132418    Wilson    Provisional   
132419    Wilson    Provisional   

                                               officename   electdate  \
0                                                Governor  2018-11-06   
1                                                Governor  2018-11-06   
2                                                Governor  2018-11-06   
3                                    United States Senate  2018-11-06   
4       United States House of Representatives District 3  2018-11-06   
...                                                   ...         ...   
132415                               United States Senate  2018-11-06   
132416  United States House of Representatives District 6  2018-11-06   
132417                       Tennessee Senate District 17  2018-11-06   
132418     Tennessee House of Representatives District 46  2018-11-06   
132419     Tennessee House of Representatives District 57  2018-11-06   

                          name        party  vote_count  
0                     Bill Lee   Republican      1011.0  
1           Tommy Ray McAnally  Independent         0.0  
2       Tracy C. Yaste Tisdale  Independent         0.0  
3             Marsha Blackburn   Republican       954.0  
4            Chuck Fleischmann   Republican      1010.0  
...                        ...          ...         ...  
132415                     NaN          NaN         NaN  
132416                     NaN          NaN         NaN  
132417                     NaN          NaN         NaN  
132418                     NaN          NaN         NaN  
132419                     NaN          NaN         NaN  

[132420 rows x 7 columns]

In [11]:
# change null values in "party" column to "no party"
tn_2018g_namemelt['party'].fillna("No party", inplace = True)
tn_2018g_namemelt

county       precinct  \
0       Anderson  Andersonville   
1       Anderson  Andersonville   
2       Anderson  Andersonville   
3       Anderson  Andersonville   
4       Anderson  Andersonville   
...          ...            ...   
132415    Wilson    Provisional   
132416    Wilson    Provisional   
132417    Wilson    Provisional   
132418    Wilson    Provisional   
132419    Wilson    Provisional   

                                               officename   electdate  \
0                                                Governor  2018-11-06   
1                                                Governor  2018-11-06   
2                                                Governor  2018-11-06   
3                                    United States Senate  2018-11-06   
4       United States House of Representatives District 3  2018-11-06   
...                                                   ...         ...   
132415                               United States Senate  2018-11-06   
132416  United States House of Representatives District 6  2018-11-06   
132417                       Tennessee Senate District 17  2018-11-06   
132418     Tennessee House of Representatives District 46  2018-11-06   
132419     Tennessee House of Representatives District 57  2018-11-06   

                          name        party  vote_count  
0                     Bill Lee   Republican      1011.0  
1           Tommy Ray McAnally  Independent         0.0  
2       Tracy C. Yaste Tisdale  Independent         0.0  
3             Marsha Blackburn   Republican       954.0  
4            Chuck Fleischmann   Republican      1010.0  
...                        ...          ...         ...  
132415                     NaN     No party         NaN  
132416                     NaN     No party         NaN  
132417                     NaN     No party         NaN  
132418                     NaN     No party         NaN  
132419                     NaN     No party         NaN  

[132420 rows x 7 columns]

In [12]:
tn_2018g_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132420 entries, 0 to 132419
Data columns (total 7 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   county      132420 non-null  object 
 1   precinct    132420 non-null  object 
 2   officename  132420 non-null  object 
 3   electdate   132420 non-null  object 
 4   name        87282 non-null   object 
 5   party       132420 non-null  object 
 6   vote_count  87282 non-null   float64
dtypes: float64(1), object(6)
memory usage: 7.1+ MB


In [13]:
# change null values in "vote_count" column to "0"
tn_2018g_namemelt['vote_count'].fillna(0)

0         1011.0
1            0.0
2            0.0
3          954.0
4         1010.0
           ...  
132415       0.0
132416       0.0
132417       0.0
132418       0.0
132419       0.0
Name: vote_count, Length: 132420, dtype: float64

In [14]:
tn_2018g_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132420 entries, 0 to 132419
Data columns (total 7 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   county      132420 non-null  object 
 1   precinct    132420 non-null  object 
 2   officename  132420 non-null  object 
 3   electdate   132420 non-null  object 
 4   name        87282 non-null   object 
 5   party       132420 non-null  object 
 6   vote_count  87282 non-null   float64
dtypes: float64(1), object(6)
memory usage: 7.1+ MB


### 87282 non-null names and 87282 non-null vote counts
### Do the nulls in "name" correlate one to one with "no party"?

In [15]:
tn_2018g_namemelt[tn_2018g_namemelt['name'].isna()]

county           precinct  \
13877   Bradley       1-3 Hopewell   
13891   Bradley  2-2 Walker Valley   
13913   Bradley      4-1 Oak Grove   
13920   Bradley   4-2 Michigan Ave   
13941   Bradley    6-1 Valley View   
...         ...                ...   
132415   Wilson        Provisional   
132416   Wilson        Provisional   
132417   Wilson        Provisional   
132418   Wilson        Provisional   
132419   Wilson        Provisional   

                                               officename   electdate name  \
13877      Tennessee House of Representatives District 22  2018-11-06  NaN   
13891      Tennessee House of Representatives District 22  2018-11-06  NaN   
13913      Tennessee House of Representatives District 22  2018-11-06  NaN   
13920      Tennessee House of Representatives District 22  2018-11-06  NaN   
13941      Tennessee House of Representatives District 22  2018-11-06  NaN   
...                                                   ...         ...  ...   
132415                               United States Senate  2018-11-06  NaN   
132416  United States House of Representatives District 6  2018-11-06  NaN   
132417                       Tennessee Senate District 17  2018-11-06  NaN   
132418     Tennessee House of Representatives District 46  2018-11-06  NaN   
132419     Tennessee House of Representatives District 57  2018-11-06  NaN   

           party  vote_count  
13877   No party         NaN  
13891   No party         NaN  
13913   No party         NaN  
13920   No party         NaN  
13941   No party         NaN  
...          ...         ...  
132415  No party         NaN  
132416  No party         NaN  
132417  No party         NaN  
132418  No party         NaN  
132419  No party         NaN  

[45138 rows x 7 columns]

In [16]:
tn_2018g_namemelt[tn_2018g_namemelt['name'].isna()].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 45138 entries, 13877 to 132419
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   county      45138 non-null  object 
 1   precinct    45138 non-null  object 
 2   officename  45138 non-null  object 
 3   electdate   45138 non-null  object 
 4   name        0 non-null      object 
 5   party       45138 non-null  object 
 6   vote_count  0 non-null      float64
dtypes: float64(1), object(6)
memory usage: 2.8+ MB


### Looks safe to drop rows with no votes tallied

In [17]:
# drop rows with no name and vote count
tn_2018g_final=tn_2018g_namemelt.dropna()

In [18]:
tn_2018g_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 87282 entries, 0 to 132413
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   county      87282 non-null  object 
 1   precinct    87282 non-null  object 
 2   officename  87282 non-null  object 
 3   electdate   87282 non-null  object 
 4   name        87282 non-null  object 
 5   party       87282 non-null  object 
 6   vote_count  87282 non-null  float64
dtypes: float64(1), object(6)
memory usage: 5.3+ MB


In [19]:
# convert vote_count data from float64 to int 32 to to drop ".0"
tn_2018g_final=tn_2018g_final.astype({'vote_count':'int32'})
tn_2018g_final

county       precinct  \
0       Anderson  Andersonville   
1       Anderson  Andersonville   
2       Anderson  Andersonville   
3       Anderson  Andersonville   
4       Anderson  Andersonville   
...          ...            ...   
132397    Wilson           25-2   
132404    Wilson       Absentee   
132405    Wilson       Absentee   
132412    Wilson    Provisional   
132413    Wilson    Provisional   

                                               officename   electdate  \
0                                                Governor  2018-11-06   
1                                                Governor  2018-11-06   
2                                                Governor  2018-11-06   
3                                    United States Senate  2018-11-06   
4       United States House of Representatives District 3  2018-11-06   
...                                                   ...         ...   
132397                                           Governor  2018-11-06   
132404                                           Governor  2018-11-06   
132405                                           Governor  2018-11-06   
132412                                           Governor  2018-11-06   
132413                                           Governor  2018-11-06   

                           name        party  vote_count  
0                      Bill Lee   Republican        1011  
1            Tommy Ray McAnally  Independent           0  
2        Tracy C. Yaste Tisdale  Independent           0  
3              Marsha Blackburn   Republican         954  
4             Chuck Fleischmann   Republican        1010  
...                         ...          ...         ...  
132397  Jeremy Allen Stephenson  Independent           0  
132404             Matthew Koch  Independent           0  
132405  Jeremy Allen Stephenson  Independent           0  
132412             Matthew Koch  Independent           0  
132413  Jeremy Allen Stephenson  Independent           0  

[87282 rows x 7 columns]

In [20]:
tn_2018g_final.dtypes

county        object
precinct      object
officename    object
electdate     object
name          object
party         object
vote_count     int32
dtype: object

In [21]:
tn_2018g_final['officename'].value_counts()

Governor                                             57652
United States Senate                                 15904
United States House of Representatives District 7     1124
United States House of Representatives District 2     1068
United States House of Representatives District 6     1064
                                                     ...  
Tennessee House of Representatives District 98          18
Tennessee House of Representatives District 88          16
Tennessee House of Representatives District 93          16
Tennessee House of Representatives District 1           11
Tennessee House of Representatives District 81           9
Name: officename, Length: 128, dtype: int64

In [22]:
tn_2018g_final['officename'].unique()

array(['Governor', 'United States Senate',
       'United States House of Representatives District 3',
       'Tennessee Senate District 5',
       'Tennessee House of Representatives District 36',
       'Tennessee House of Representatives District 33',
       'United States House of Representatives District 4',
       'Tennessee House of Representatives District 62',
       'United States House of Representatives District 7',
       'Tennessee House of Representatives District 75',
       'United States House of Representatives District 8',
       'Tennessee Senate District 15',
       'Tennessee House of Representatives District 31',
       'United States House of Representatives District 2',
       'Tennessee Senate District 2 (unexpired term)',
       'Tennessee House of Representatives District 8',
       'Tennessee House of Representatives District 20',
       'Tennessee House of Representatives District 24',
       'Tennessee Senate District 9',
       'Tennessee House of Repre

In [23]:
tn_2018g_final['vote_count'].value_counts()

0       41700
1       12006
2        5343
3        2925
4        1791
        ...  
1708        1
2254        1
1900        1
2126        1
2047        1
Name: vote_count, Length: 2155, dtype: int64

### 41700 rows have a vote count of 0. Drop those rows

In [24]:
tn_2018g_final=tn_2018g_final[tn_2018g_final['vote_count'] != 0]
tn_2018g_final=tn_2018g_final.reset_index(drop=True)
tn_2018g_final.head()

county       precinct                                         officename  \
0  Anderson  Andersonville                                           Governor   
1  Anderson  Andersonville                               United States Senate   
2  Anderson  Andersonville  United States House of Representatives District 3   
3  Anderson  Andersonville                        Tennessee Senate District 5   
4  Anderson  Andersonville     Tennessee House of Representatives District 36   

    electdate               name       party  vote_count  
0  2018-11-06           Bill Lee  Republican        1011  
1  2018-11-06   Marsha Blackburn  Republican         954  
2  2018-11-06  Chuck Fleischmann  Republican        1010  
3  2018-11-06      Randy McNally  Republican        1051  
4  2018-11-06   Dennis H. Powers  Republican         996

In [25]:
# confirm no 0's in vote count
tn_2018g_final['vote_count'].value_counts()

1        12006
2         5343
3         2925
4         1791
5         1207
         ...  
11752        1
1389         1
9545         1
3340         1
2049         1
Name: vote_count, Length: 2154, dtype: int64

In [29]:
# create data frame tn_2018_uss_int for US Senate election
tn_2018_uss_int=tn_2018g_final[tn_2018g_final['officename'].str.match('United States Senate')]
tn_2018_uss_int=tn_2018_uss_int.reset_index(drop=True)
tn_2018_uss_int

county       precinct            officename   electdate  \
0      Anderson  Andersonville  United States Senate  2018-11-06   
1      Anderson     Briceville  United States Senate  2018-11-06   
2      Anderson       Bull Run  United States Senate  2018-11-06   
3      Anderson        Clinton  United States Senate  2018-11-06   
4      Anderson   Clinton High  United States Senate  2018-11-06   
...         ...            ...                   ...         ...   
12454    Wilson           23-1  United States Senate  2018-11-06   
12455    Wilson           24-1  United States Senate  2018-11-06   
12456    Wilson           25-1  United States Senate  2018-11-06   
12457    Wilson           25-2  United States Senate  2018-11-06   
12458    Wilson       Absentee  United States Senate  2018-11-06   

                   name        party  vote_count  
0      Marsha Blackburn   Republican         954  
1      Marsha Blackburn   Republican         176  
2      Marsha Blackburn   Republican        1074  
3      Marsha Blackburn   Republican        1032  
4      Marsha Blackburn   Republican         618  
...                 ...          ...         ...  
12454      Kris L. Todd  Independent           8  
12455      Kris L. Todd  Independent          11  
12456      Kris L. Todd  Independent          10  
12457      Kris L. Todd  Independent          12  
12458      Kris L. Todd  Independent           3  

[12459 rows x 7 columns]

In [30]:
tn_2018_uss_int.head(20)

county        precinct            officename   electdate  \
0   Anderson   Andersonville  United States Senate  2018-11-06   
1   Anderson      Briceville  United States Senate  2018-11-06   
2   Anderson        Bull Run  United States Senate  2018-11-06   
3   Anderson         Clinton  United States Senate  2018-11-06   
4   Anderson    Clinton High  United States Senate  2018-11-06   
5   Anderson         Claxton  United States Senate  2018-11-06   
6   Anderson  Clinton Middle  United States Senate  2018-11-06   
7   Anderson    Dutch Valley  United States Senate  2018-11-06   
8   Anderson    Emory Valley  United States Senate  2018-11-06   
9   Anderson        Fairview  United States Senate  2018-11-06   
10  Anderson     Glen Alpine  United States Senate  2018-11-06   
11  Anderson        Glenwood  United States Senate  2018-11-06   
12  Anderson   Hendrix Creek  United States Senate  2018-11-06   
13  Anderson   Highland View  United States Senate  2018-11-06   
14  Anderson    Lake City MS  United States Senate  2018-11-06   
15  Anderson          Marlow  United States Senate  2018-11-06   
16  Anderson   North Clinton  United States Senate  2018-11-06   
17  Anderson          Norris  United States Senate  2018-11-06   
18  Anderson         Norwood  United States Senate  2018-11-06   
19  Anderson       Oak Ridge  United States Senate  2018-11-06   

                name       party  vote_count  
0   Marsha Blackburn  Republican         954  
1   Marsha Blackburn  Republican         176  
2   Marsha Blackburn  Republican        1074  
3   Marsha Blackburn  Republican        1032  
4   Marsha Blackburn  Republican         618  
5   Marsha Blackburn  Republican         991  
6   Marsha Blackburn  Republican         336  
7   Marsha Blackburn  Republican         196  
8   Marsha Blackburn  Republican         835  
9   Marsha Blackburn  Republican         452  
10  Marsha Blackburn  Republican         851  
11  Marsha Blackburn  Republican         586  
12  Marsha Blackburn  Republican         770  
13  Marsha Blackburn  Republican         241  
14  Marsha Blackburn  Republican         839  
15  Marsha Blackburn  Republican         368  
16  Marsha Blackburn  Republican         622  
17  Marsha Blackburn  Republican         321  
18  Marsha Blackburn  Republican         698  
19  Marsha Blackburn  Republican         234

### Add column that has total vote count by count for use in calculating county vote percentages

In [31]:
# group county precincts into county per candidate, in descending order of votes within county
tn_2018_uss_int=tn_2018_uss_int.groupby(['electdate','officename','county','name','party'], as_index=False)['vote_count'].sum()
tn_2018_uss=tn_2018_uss_int.sort_values(by=['county','vote_count'],ascending=False).reset_index(drop=True)
tn_2018_uss.head(30)

electdate            officename      county               name  \
0   2018-11-06  United States Senate      Wilson   Marsha Blackburn   
1   2018-11-06  United States Senate      Wilson      Phil Bredesen   
2   2018-11-06  United States Senate      Wilson       Kris L. Todd   
3   2018-11-06  United States Senate      Wilson    Trudy A. Austin   
4   2018-11-06  United States Senate      Wilson          Dean Hill   
5   2018-11-06  United States Senate      Wilson        John Carico   
6   2018-11-06  United States Senate      Wilson    Breton Phillips   
7   2018-11-06  United States Senate      Wilson  Kevin Lee McCants   
8   2018-11-06  United States Senate  Williamson   Marsha Blackburn   
9   2018-11-06  United States Senate  Williamson      Phil Bredesen   
10  2018-11-06  United States Senate  Williamson    Trudy A. Austin   
11  2018-11-06  United States Senate  Williamson          Dean Hill   
12  2018-11-06  United States Senate  Williamson       Kris L. Todd   
13  2018-11-06  United States Senate  Williamson        John Carico   
14  2018-11-06  United States Senate  Williamson    Breton Phillips   
15  2018-11-06  United States Senate  Williamson  Kevin Lee McCants   
16  2018-11-06  United States Senate       White   Marsha Blackburn   
17  2018-11-06  United States Senate       White      Phil Bredesen   
18  2018-11-06  United States Senate       White    Trudy A. Austin   
19  2018-11-06  United States Senate       White          Dean Hill   
20  2018-11-06  United States Senate       White        John Carico   
21  2018-11-06  United States Senate       White       Kris L. Todd   
22  2018-11-06  United States Senate       White    Breton Phillips   
23  2018-11-06  United States Senate       White  Kevin Lee McCants   
24  2018-11-06  United States Senate     Weakley   Marsha Blackburn   
25  2018-11-06  United States Senate     Weakley      Phil Bredesen   
26  2018-11-06  United States Senate     Weakley          Dean Hill   
27  2018-11-06  United States Senate     Weakley    Trudy A. Austin   
28  2018-11-06  United States Senate     Weakley       Kris L. Todd   
29  2018-11-06  United States Senate     Weakley  Kevin Lee McCants   

          party  vote_count  
0    Republican       32810  
1    Democratic       19447  
2   Independent         289  
3   Independent         220  
4   Independent         139  
5   Independent          61  
6   Independent          57  
7   Independent          31  
8    Republican       62039  
9    Democratic       42611  
10  Independent         264  
11  Independent         153  
12  Independent         126  
13  Independent          82  
14  Independent          54  
15  Independent          40  
16   Republican        5979  
17   Democratic        2293  
18  Independent          59  
19  Independent          41  
20  Independent          19  
21  Independent          18  
22  Independent           9  
23  Independent           5  
24   Republican        6704  
25   Democratic        3047  
26  Independent          40  
27  Independent          33  
28  Independent          30  
29  Independent          23

In [32]:
# create new dataframe "county_vote" to use to find sum of votes per county
county_vote=tn_2018_uss.loc[0:,['county','vote_count']]
county_vote

county  vote_count
0      Wilson       32810
1      Wilson       19447
2      Wilson         289
3      Wilson         220
4      Wilson         139
..        ...         ...
753  Anderson         150
754  Anderson          48
755  Anderson          38
756  Anderson          35
757  Anderson          23

[758 rows x 2 columns]

In [33]:
#sum county votes
county_vote_sum=county_vote.groupby(['county'])['vote_count'].sum()
county_vote_sum

county
Anderson       26008
Bedford        13033
Benton          5298
Bledsoe         4076
Blount         46782
               ...  
Wayne           4736
Weakley         9901
White           8423
Williamson    105369
Wilson         53054
Name: vote_count, Length: 95, dtype: int32

In [34]:
# use "county_vote_sum" to add column "sum_county_vote" to tn_2018_uss correlating with value in each row
tn_2018_uss['sum_county_vote']=tn_2018_uss['county'].map(county_vote_sum)
tn_2018_uss.head(20)

electdate            officename      county               name  \
0   2018-11-06  United States Senate      Wilson   Marsha Blackburn   
1   2018-11-06  United States Senate      Wilson      Phil Bredesen   
2   2018-11-06  United States Senate      Wilson       Kris L. Todd   
3   2018-11-06  United States Senate      Wilson    Trudy A. Austin   
4   2018-11-06  United States Senate      Wilson          Dean Hill   
5   2018-11-06  United States Senate      Wilson        John Carico   
6   2018-11-06  United States Senate      Wilson    Breton Phillips   
7   2018-11-06  United States Senate      Wilson  Kevin Lee McCants   
8   2018-11-06  United States Senate  Williamson   Marsha Blackburn   
9   2018-11-06  United States Senate  Williamson      Phil Bredesen   
10  2018-11-06  United States Senate  Williamson    Trudy A. Austin   
11  2018-11-06  United States Senate  Williamson          Dean Hill   
12  2018-11-06  United States Senate  Williamson       Kris L. Todd   
13  2018-11-06  United States Senate  Williamson        John Carico   
14  2018-11-06  United States Senate  Williamson    Breton Phillips   
15  2018-11-06  United States Senate  Williamson  Kevin Lee McCants   
16  2018-11-06  United States Senate       White   Marsha Blackburn   
17  2018-11-06  United States Senate       White      Phil Bredesen   
18  2018-11-06  United States Senate       White    Trudy A. Austin   
19  2018-11-06  United States Senate       White          Dean Hill   

          party  vote_count  sum_county_vote  
0    Republican       32810            53054  
1    Democratic       19447            53054  
2   Independent         289            53054  
3   Independent         220            53054  
4   Independent         139            53054  
5   Independent          61            53054  
6   Independent          57            53054  
7   Independent          31            53054  
8    Republican       62039           105369  
9    Democratic       42611           105369  
10  Independent         264           105369  
11  Independent         153           105369  
12  Independent         126           105369  
13  Independent          82           105369  
14  Independent          54           105369  
15  Independent          40           105369  
16   Republican        5979             8423  
17   Democratic        2293             8423  
18  Independent          59             8423  
19  Independent          41             8423

In [35]:
# create "pct_county_vote" column for tn_2018_uss
tn_2018_uss['pct_county_vote']=round(tn_2018_uss['vote_count']/tn_2018_uss['sum_county_vote']*100,2)
tn_2018_uss

electdate            officename    county               name  \
0    2018-11-06  United States Senate    Wilson   Marsha Blackburn   
1    2018-11-06  United States Senate    Wilson      Phil Bredesen   
2    2018-11-06  United States Senate    Wilson       Kris L. Todd   
3    2018-11-06  United States Senate    Wilson    Trudy A. Austin   
4    2018-11-06  United States Senate    Wilson          Dean Hill   
..          ...                   ...       ...                ...   
753  2018-11-06  United States Senate  Anderson    Trudy A. Austin   
754  2018-11-06  United States Senate  Anderson       Kris L. Todd   
755  2018-11-06  United States Senate  Anderson        John Carico   
756  2018-11-06  United States Senate  Anderson    Breton Phillips   
757  2018-11-06  United States Senate  Anderson  Kevin Lee McCants   

           party  vote_count  sum_county_vote  pct_county_vote  
0     Republican       32810            53054            61.84  
1     Democratic       19447            53054            36.66  
2    Independent         289            53054             0.54  
3    Independent         220            53054             0.41  
4    Independent         139            53054             0.26  
..           ...         ...              ...              ...  
753  Independent         150            26008             0.58  
754  Independent          48            26008             0.18  
755  Independent          38            26008             0.15  
756  Independent          35            26008             0.13  
757  Independent          23            26008             0.09  

[758 rows x 8 columns]

In [37]:
# how many parties?
tn_2018_uss['party'].value_counts()

Independent    568
Republican      95
Democratic      95
Name: party, dtype: int64

### Create dataframe "tn_2018_uss_pct_county"  with percent values per party per county

In [38]:
# create new dataframe with reduced columns values only
tn_2018_uss_pct_county=tn_2018_uss.loc[0:,['electdate','officename','county','party','pct_county_vote']]
tn_2018_uss_pct_county

electdate            officename    county        party  pct_county_vote
0    2018-11-06  United States Senate    Wilson   Republican            61.84
1    2018-11-06  United States Senate    Wilson   Democratic            36.66
2    2018-11-06  United States Senate    Wilson  Independent             0.54
3    2018-11-06  United States Senate    Wilson  Independent             0.41
4    2018-11-06  United States Senate    Wilson  Independent             0.26
..          ...                   ...       ...          ...              ...
753  2018-11-06  United States Senate  Anderson  Independent             0.58
754  2018-11-06  United States Senate  Anderson  Independent             0.18
755  2018-11-06  United States Senate  Anderson  Independent             0.15
756  2018-11-06  United States Senate  Anderson  Independent             0.13
757  2018-11-06  United States Senate  Anderson  Independent             0.09

[758 rows x 5 columns]

In [39]:
# create new dataframe with Republican values only
tn_2018_uss_rep_pct=tn_2018_uss_pct_county[tn_2018_uss_pct_county['party']== "Republican"].reset_index(drop=True)
tn_2018_uss_rep_pct

electdate            officename      county       party  pct_county_vote
0   2018-11-06  United States Senate      Wilson  Republican            61.84
1   2018-11-06  United States Senate  Williamson  Republican            58.88
2   2018-11-06  United States Senate       White  Republican            70.98
3   2018-11-06  United States Senate     Weakley  Republican            67.71
4   2018-11-06  United States Senate       Wayne  Republican            79.96
..         ...                   ...         ...         ...              ...
90  2018-11-06  United States Senate      Blount  Republican            63.97
91  2018-11-06  United States Senate     Bledsoe  Republican            71.34
92  2018-11-06  United States Senate      Benton  Republican            65.86
93  2018-11-06  United States Senate     Bedford  Republican            68.61
94  2018-11-06  United States Senate    Anderson  Republican            56.60

[95 rows x 5 columns]

In [40]:
# rename 'pct_county_vote' to 'rep_pct' and drop "party" column to prepare for merge 
tn_2018_uss_rep_pct=tn_2018_uss_rep_pct.rename(columns={'pct_county_vote':'Republican'},inplace=False).drop(['party'], axis=1)
tn_2018_uss_rep_pct

electdate            officename      county  Republican
0   2018-11-06  United States Senate      Wilson       61.84
1   2018-11-06  United States Senate  Williamson       58.88
2   2018-11-06  United States Senate       White       70.98
3   2018-11-06  United States Senate     Weakley       67.71
4   2018-11-06  United States Senate       Wayne       79.96
..         ...                   ...         ...         ...
90  2018-11-06  United States Senate      Blount       63.97
91  2018-11-06  United States Senate     Bledsoe       71.34
92  2018-11-06  United States Senate      Benton       65.86
93  2018-11-06  United States Senate     Bedford       68.61
94  2018-11-06  United States Senate    Anderson       56.60

[95 rows x 4 columns]

In [41]:
# create new dataframe with Democratic values only
tn_2018_uss_dem_pct=tn_2018_uss_pct_county[tn_2018_uss_pct_county['party']== "Democratic"].reset_index(drop=True)
tn_2018_uss_dem_pct

electdate            officename      county       party  pct_county_vote
0   2018-11-06  United States Senate      Wilson  Democratic            36.66
1   2018-11-06  United States Senate  Williamson  Democratic            40.44
2   2018-11-06  United States Senate       White  Democratic            27.22
3   2018-11-06  United States Senate     Weakley  Democratic            30.77
4   2018-11-06  United States Senate       Wayne  Democratic            19.13
..         ...                   ...         ...         ...              ...
90  2018-11-06  United States Senate      Blount  Democratic            34.29
91  2018-11-06  United States Senate     Bledsoe  Democratic            26.89
92  2018-11-06  United States Senate      Benton  Democratic            32.50
93  2018-11-06  United States Senate     Bedford  Democratic            30.15
94  2018-11-06  United States Senate    Anderson  Democratic            41.59

[95 rows x 5 columns]

In [42]:
# rename 'pct_county_vote' to 'dem_pct' and drop "party" to prepare for merge 
tn_2018_uss_dem_pct=tn_2018_uss_dem_pct.rename(columns={'pct_county_vote':'Democratic'},inplace=False).drop(['party'], axis=1)
tn_2018_uss_dem_pct

electdate            officename      county  Democratic
0   2018-11-06  United States Senate      Wilson       36.66
1   2018-11-06  United States Senate  Williamson       40.44
2   2018-11-06  United States Senate       White       27.22
3   2018-11-06  United States Senate     Weakley       30.77
4   2018-11-06  United States Senate       Wayne       19.13
..         ...                   ...         ...         ...
90  2018-11-06  United States Senate      Blount       34.29
91  2018-11-06  United States Senate     Bledsoe       26.89
92  2018-11-06  United States Senate      Benton       32.50
93  2018-11-06  United States Senate     Bedford       30.15
94  2018-11-06  United States Senate    Anderson       41.59

[95 rows x 4 columns]

In [44]:
# create new dataframe with non Republican or Democratic values only
tn_2018_uss_other_pct=tn_2018_uss_pct_county[(tn_2018_uss_pct_county['party']!="Republican")&(tn_2018_uss_pct_county['party']!="Democratic")].reset_index(drop=True)
tn_2018_uss_other_pct

electdate            officename    county        party  pct_county_vote
0    2018-11-06  United States Senate    Wilson  Independent             0.54
1    2018-11-06  United States Senate    Wilson  Independent             0.41
2    2018-11-06  United States Senate    Wilson  Independent             0.26
3    2018-11-06  United States Senate    Wilson  Independent             0.11
4    2018-11-06  United States Senate    Wilson  Independent             0.11
..          ...                   ...       ...          ...              ...
563  2018-11-06  United States Senate  Anderson  Independent             0.58
564  2018-11-06  United States Senate  Anderson  Independent             0.18
565  2018-11-06  United States Senate  Anderson  Independent             0.15
566  2018-11-06  United States Senate  Anderson  Independent             0.13
567  2018-11-06  United States Senate  Anderson  Independent             0.09

[568 rows x 5 columns]

In [45]:
# sum by county
tn_2018_uss_other_pct=tn_2018_uss_other_pct.groupby(['county'],as_index=False)['pct_county_vote'].sum()
tn_2018_uss_other_pct

county  pct_county_vote
0     Anderson             1.81
1      Bedford             1.25
2       Benton             1.64
3      Bledsoe             1.77
4       Blount             1.73
..         ...              ...
90       Wayne             0.90
91     Weakley             1.50
92       White             1.80
93  Williamson             0.69
94      Wilson             1.49

[95 rows x 2 columns]

In [46]:
# rename 'pct_county_vote' to 'other_pct' and drop "party" to prepare for merge 
tn_2018_uss_other_pct=tn_2018_uss_other_pct.rename(columns={'pct_county_vote':'Other'})
tn_2018_uss_other_pct

county  Other
0     Anderson   1.81
1      Bedford   1.25
2       Benton   1.64
3      Bledsoe   1.77
4       Blount   1.73
..         ...    ...
90       Wayne   0.90
91     Weakley   1.50
92       White   1.80
93  Williamson   0.69
94      Wilson   1.49

[95 rows x 2 columns]

In [47]:
# join tn_2018_uss_rep, tn_2018_uss_dem_pct and tn_2018__other_pct to form tn_2018_uss_county_pct
dem=tn_2018_uss_dem_pct['Democratic']
other=tn_2018_uss_other_pct['Other']
tn_2018_uss_county_pct=tn_2018_uss_rep_pct.join(dem).join(other)
tn_2018_uss_county_pct

electdate            officename      county  Republican  Democratic  \
0   2018-11-06  United States Senate      Wilson       61.84       36.66   
1   2018-11-06  United States Senate  Williamson       58.88       40.44   
2   2018-11-06  United States Senate       White       70.98       27.22   
3   2018-11-06  United States Senate     Weakley       67.71       30.77   
4   2018-11-06  United States Senate       Wayne       79.96       19.13   
..         ...                   ...         ...         ...         ...   
90  2018-11-06  United States Senate      Blount       63.97       34.29   
91  2018-11-06  United States Senate     Bledsoe       71.34       26.89   
92  2018-11-06  United States Senate      Benton       65.86       32.50   
93  2018-11-06  United States Senate     Bedford       68.61       30.15   
94  2018-11-06  United States Senate    Anderson       56.60       41.59   

    Other  
0    1.81  
1    1.25  
2    1.64  
3    1.77  
4    1.73  
..    ...  
90   0.90  
91   1.50  
92   1.80  
93   0.69  
94   1.49  

[95 rows x 6 columns]

### Add column for Republican margin over Democrat for heat mapping in Tableau

In [48]:
tn_2018_uss_county_pct['Rep_Margin']= tn_2018_uss_county_pct['Republican']-tn_2018_uss_county_pct['Democratic']
tn_2018_uss_county_pct

electdate            officename      county  Republican  Democratic  \
0   2018-11-06  United States Senate      Wilson       61.84       36.66   
1   2018-11-06  United States Senate  Williamson       58.88       40.44   
2   2018-11-06  United States Senate       White       70.98       27.22   
3   2018-11-06  United States Senate     Weakley       67.71       30.77   
4   2018-11-06  United States Senate       Wayne       79.96       19.13   
..         ...                   ...         ...         ...         ...   
90  2018-11-06  United States Senate      Blount       63.97       34.29   
91  2018-11-06  United States Senate     Bledsoe       71.34       26.89   
92  2018-11-06  United States Senate      Benton       65.86       32.50   
93  2018-11-06  United States Senate     Bedford       68.61       30.15   
94  2018-11-06  United States Senate    Anderson       56.60       41.59   

    Other  Rep_Margin  
0    1.81       25.18  
1    1.25       18.44  
2    1.64       43.76  
3    1.77       36.94  
4    1.73       60.83  
..    ...         ...  
90   0.90       29.68  
91   1.50       44.45  
92   1.80       33.36  
93   0.69       38.46  
94   1.49       15.01  

[95 rows x 7 columns]

In [49]:
#read TN_county_index_shape.csv to add index from TN Counties shape file for joining in Tableau
county_index=pd.read_csv('../data/TN_county_index_shape.csv')
county_index

county  INDEX_NO
0     Anderson        79
1      Bedford        26
2       Benton        41
3      Bledsoe        31
4       Blount        52
..         ...       ...
90       Wayne         6
91     Weakley        56
92       White        67
93  Williamson        44
94      Wilson        65

[95 rows x 2 columns]

In [50]:
# join with tn_2018_uss_county_pct
tn_2018_uss_county_pct=tn_2018_uss_county_pct.merge(county_index,on="county")
tn_2018_uss_county_pct.head(50)

electdate            officename      county  Republican  Democratic  \
0   2018-11-06  United States Senate      Wilson       61.84       36.66   
1   2018-11-06  United States Senate  Williamson       58.88       40.44   
2   2018-11-06  United States Senate       White       70.98       27.22   
3   2018-11-06  United States Senate     Weakley       67.71       30.77   
4   2018-11-06  United States Senate       Wayne       79.96       19.13   
5   2018-11-06  United States Senate  Washington       63.08       35.94   
6   2018-11-06  United States Senate      Warren       61.46       36.78   
7   2018-11-06  United States Senate   Van Buren       65.75       32.22   
8   2018-11-06  United States Senate       Union       71.52       26.43   
9   2018-11-06  United States Senate      Unicoi       74.19       24.75   
10  2018-11-06  United States Senate   Trousdale       60.35       38.13   
11  2018-11-06  United States Senate      Tipton       69.05       29.01   
12  2018-11-06  United States Senate      Sumner       62.68       36.24   
13  2018-11-06  United States Senate    Sullivan       71.12       27.86   
14  2018-11-06  United States Senate     Stewart       65.36       33.19   
15  2018-11-06  United States Senate       Smith       66.25       32.40   
16  2018-11-06  United States Senate      Shelby       33.14       65.56   
17  2018-11-06  United States Senate      Sevier       71.60       25.82   
18  2018-11-06  United States Senate  Sequatchie       71.04       27.25   
19  2018-11-06  United States Senate       Scott       76.93       21.38   
20  2018-11-06  United States Senate  Rutherford       52.72       46.10   
21  2018-11-06  United States Senate   Robertson       64.30       34.47   
22  2018-11-06  United States Senate       Roane       65.81       32.28   
23  2018-11-06  United States Senate        Rhea       72.95       25.19   
24  2018-11-06  United States Senate      Putnam       62.59       35.96   
25  2018-11-06  United States Senate        Polk       70.92       27.24   
26  2018-11-06  United States Senate     Pickett       70.44       28.46   
27  2018-11-06  United States Senate       Perry       67.60       29.77   
28  2018-11-06  United States Senate     Overton       64.49       34.08   
29  2018-11-06  United States Senate       Obion       74.46       24.33   
30  2018-11-06  United States Senate      Morgan       72.39       25.46   
31  2018-11-06  United States Senate       Moore       74.12       24.40   
32  2018-11-06  United States Senate  Montgomery       49.98       47.81   
33  2018-11-06  United States Senate      Monroe       71.05       26.54   
34  2018-11-06  United States Senate       Meigs       70.99       27.40   
35  2018-11-06  United States Senate     McNairy       73.36       24.87   
36  2018-11-06  United States Senate      McMinn       72.28       26.06   
37  2018-11-06  United States Senate       Maury       60.10       38.58   
38  2018-11-06  United States Senate    Marshall       65.16       33.66   
39  2018-11-06  United States Senate      Marion       63.65       34.51   
40  2018-11-06  United States Senate     Madison       52.05       46.54   
41  2018-11-06  United States Senate       Macon       75.11       23.36   
42  2018-11-06  United States Senate      Loudon       68.66       29.49   
43  2018-11-06  United States Senate     Lincoln       75.40       23.12   
44  2018-11-06  United States Senate       Lewis       69.96       29.03   
45  2018-11-06  United States Senate    Lawrence       72.22       26.52   
46  2018-11-06  United States Senate  Lauderdale       56.14       42.12   
47  2018-11-06  United States Senate        Lake       64.09       33.77   
48  2018-11-06  United States Senate        Knox       50.80       47.71   
49  2018-11-06  United States Senate     Johnson       76.64       22.06   

    Other  Rep_Margin  INDEX_NO  
0    1.81       25.18        65  
1    1.25       18.44        44  
2    1.64       43.76     

In [51]:
tn_2018_uss_county_pct.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 95 entries, 0 to 94
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   electdate   95 non-null     object 
 1   officename  95 non-null     object 
 2   county      95 non-null     object 
 3   Republican  95 non-null     float64
 4   Democratic  95 non-null     float64
 5   Other       95 non-null     float64
 6   Rep_Margin  95 non-null     float64
 7   INDEX_NO    95 non-null     int64  
dtypes: float64(4), int64(1), object(3)
memory usage: 6.7+ KB


In [52]:
# covert "electdate" to datetime from object
tn_2018_uss_county_pct['electdate']=pd.to_datetime(tn_2018_uss_county_pct['electdate'],format='%Y-%m-%d')
tn_2018_uss_county_pct

electdate            officename      county  Republican  Democratic  \
0  2018-11-06  United States Senate      Wilson       61.84       36.66   
1  2018-11-06  United States Senate  Williamson       58.88       40.44   
2  2018-11-06  United States Senate       White       70.98       27.22   
3  2018-11-06  United States Senate     Weakley       67.71       30.77   
4  2018-11-06  United States Senate       Wayne       79.96       19.13   
..        ...                   ...         ...         ...         ...   
90 2018-11-06  United States Senate      Blount       63.97       34.29   
91 2018-11-06  United States Senate     Bledsoe       71.34       26.89   
92 2018-11-06  United States Senate      Benton       65.86       32.50   
93 2018-11-06  United States Senate     Bedford       68.61       30.15   
94 2018-11-06  United States Senate    Anderson       56.60       41.59   

    Other  Rep_Margin  INDEX_NO  
0    1.81       25.18        65  
1    1.25       18.44        44  
2    1.64       43.76        67  
3    1.77       36.94        56  
4    1.73       60.83         6  
..    ...         ...       ...  
90   0.90       29.68        52  
91   1.50       44.45        31  
92   1.80       33.36        41  
93   0.69       38.46        26  
94   1.49       15.01        79  

[95 rows x 8 columns]

In [54]:
tn_2018_uss_county_pct.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 95 entries, 0 to 94
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   electdate   95 non-null     datetime64[ns]
 1   officename  95 non-null     object        
 2   county      95 non-null     object        
 3   Republican  95 non-null     float64       
 4   Democratic  95 non-null     float64       
 5   Other       95 non-null     float64       
 6   Rep_Margin  95 non-null     float64       
 7   INDEX_NO    95 non-null     int64         
dtypes: datetime64[ns](1), float64(4), int64(1), object(2)
memory usage: 6.7+ KB


In [55]:
tn_2018_uss_county_pct.to_excel(r'../data/2018_election/tn_2018_uss_county_pct.xlsx',index=False)